docs/multiple_image_support_design_proposal.md

```diff
@@ -0,0 +1,549 @@
+
+`Note: Might be this double linked dependency handling can be replaced with
+something simpler, i.e.: single linked. That case should be a main image, which
+has no dependency and other images can depend only on previous image in the
+chain.`
+
+# How to include dependency entries?
+In case of a single vendor, who owns all the images but still want to do
+independent updates for example to save power, to including dependency entries
+as a TLV is pretty easy. Only the scripts should be modified.
+
+In case of multi vendor scenario when the device maker/owner only owns the main
+application and other SW images is provided by other party, then it might, not
+straight forward to ask the other party to include my main application's version
+to his image as a TLV entry. Might be TLV entries must be extended by device
+owner and that part should be signed his own key?
+
+# Integrity protection of manifest
+Currently the image manifest is not needed to be part of the integrity protected
+section. There is no security vulnerability that could be exploited by modifying
+it. It contains these TLV entries:
+  - Image hash
+  - Key hash
+  - Signature (hash based)
+
+```
+  +--------------------+
+  | Header             |   <- struct image_header
+  +--------------------+
+  | Payload            |
+  +--------------------+
+  | Manifest           |
+  | +----------------+ |
+  | | Manifest header| |   <- struct image_tlv_info
+A | +----------------+ |
+  | | SHA256 hash    | |   <- hash of header and payload (struct image_tlv)
+B | +----------------+ |
+  | | Keyhash        | |   <- indicates which public key for sig (struct image_tlv)
+  | +----------------+ |
+  | | Signature      | |   <- signature from A - B(struct image_tlv), only hash
+  | +----------------+ |
+  +--------------------+
+```
+
+But if manifest will be expanded with more info, then manifest itself required
+to be integrity and authenticity protected. The easiest solution described
+below, it was proposed by David Brown. There were alternatives. I chose, which
+has less code impact.
+
+```
+  +--------------------+
+  | Header             |   <- struct image_header
+  +--------------------+
+  | Payload            |
+  +--------------------+
+  | Manifest           |
+  | +----------------+ |
+  | | Manifest header| |   <- struct image_tlv_info
+A | +----------------+ |
+  | | SHA256 hash    | |   <- hash of header and payload (struct image_tlv)
+  | +----------------+ |
+  | | Dependency     | |   <- Dependency entry(struct image_tlv)
+B | +----------------+ |
+  | | Keyhash        | |   <- indicates which public key for sig (struct image_tlv)
+  | +----------------+ |
+  | | Signature      | |   <- signature from A - B(struct image_tlv)
+  | +----------------+ |
+  +--------------------+
+```
+
+# Flash partitioning
+In order to support independent update of multiple images the flash must be
+partitioned further to arrange two slot for each image.
+```
++--------------------+
```

Comment:
Just out of curiosity, would the following be infeasible in any real-world case?

    +--------------------+
    | MCUBoot            |
    +--------------------+
          ~~~~~              <- memory might be not contiguous
    +--------------------+
    | Image 0            |
    | primary   area     |
    +--------------------+
          ~~~~~              <- memory might be not contiguous
    +--------------------+
    | Image 1            |
    | primary   area     |
    +--------------------+
          ~~~~~              <- memory might be not contiguous
    +--------------------+
    | Image 0            |
    | secondary area     |
    +--------------------+
    | Image 1            |
    | secondary area     |
    +--------------------+
    | Scratch            |
    +--------------------+

Reply:
I'm not an expert in flash technology, but considering the following I do not see an issue:
- If flash controller have access to all memory are
- If size of image 0/1  primary and secondary area has the same size
- If sector layout of image 0/1 primary and secondary area is the same

Reply:
> If sector layout of image 0/1 primary and secondary area is the same.

By a sector, do you mean the smallest area that can be erased?

What sort of weird flash layouts actually exist out there?

Reply:
Yes use sector in this context: smallest erasable area.
Some STM32 flash has mixed layout not all sector has the same size. There are big and small sector on the flash.

Reply:
Ok, thanks for the explanation.